nova-core/nlp/nlp.py

```diff
@@ -1,5 +1,69 @@
-def speech_to_text(input_: any) -> str:
-    return input_
+import io
+from threading import Semaphore
 
-def text_to_speech(input_: str) -> any:
-    return input_
\ No newline at end of file
+import gtts
+import speech_recognition as sr
+from playsound import playsound
+
+UNKNOWN_VALUE_ERROR = sr.UnknownValueError
+
+class SpeechRecognition:
+    '''Singleton implmentation of Speech Recognition utility functions'''
+
+    _instance = None
+
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = super(SpeechRecognition, cls).__new__(cls)
+            cls._instance._initialization_routine()
+        return cls._instance
+
+    def _initialization_routine(self) -> None:
+        self.recognizer = sr.Recognizer()
+        # self.tts_engine = pyttsx3.init()
+        self._mutex = Semaphore()
+        # pygame.mixer.init()
+
+    # def set_voice(self, voice_index: int) -> None:
+    #     voices = self.tts_engine.getProperty('voices')
+    #     self.tts_engine.setProperty('voice', voices[voice_index].id)
+
+    def speech_to_text(self, input_: str) -> str:
+        '''Speech to text which accepts str'''
+        return input_
+    
+    def speech_to_text(self, timeout=None) -> str:
+        text: str = ''
+        with sr.Microphone() as source:
+            self._mutex.acquire()
+            try:
+                self.recognizer.adjust_for_ambient_noise(source, duration=3)
+                print('[NLP] Listening')
```

Comment:
Line 40 isn't actually listening, its just adjusting so it can filter out background noise.  Lmk if thats not what you meant.  Also this would probably be an LED turning on/off on a physical device for context (more for logging on CLI)